### PR TITLE
Risk condition for server listen

### PR DESCRIPTION
**Pull request by zxDiscovery**
Fix https://github.com/knative/pkg/issues/417

### CODE REVIEW

File: webhook/webhook.go [hunk 1]
```diff
@@ -281,6 +283,19 @@ func (ac *AdmissionController) Run(stop <-chan struct{}) error {
...
+	go func() {
+		if err := server.ListenAndServeTLS("", ""); err != nil {
+			logger.Errorw("ListenAndServeTLS for admission webhook returned error", zap.Error(err))
+			close(serverBootstrapErrCh)
+		}
+		wg.Done()
+		logger.Info("server ready to listen")
+	}()
+
+	wg.Wait()
```
Comment (mattmoor) [inline webhook/webhook.go]: This is equivalent to not doing this in a go-routine, which I believe is incorrect given the examples:
https://golang.org/pkg/net/http/#ListenAndServe

File: webhook/webhook.go [hunk 2]
```diff
@@ -281,6 +283,19 @@ func (ac *AdmissionController) Run(stop <-chan struct{}) error {
...
 	}
 
+	serverBootstrapErrCh := make(chan struct{})
+	wg.Add(1)
+	go func() {
+		if err := server.ListenAndServeTLS("", ""); err != nil {
+			logger.Errorw("ListenAndServeTLS for admission webhook returned error", zap.Error(err))
+			close(serverBootstrapErrCh)
+		}
+		wg.Done()
```
Comment (vagababov) [inline webhook/webhook.go]: always (unless strong reason not to exists) do `defer wg.Done()`

File: webhook/webhook.go (hunk 1 above)
Comment (vagababov) [reply webhook/webhook.go]: +1 the way it's written this `wg` won't be ready until after the server starts. You need one more goroutine to probe the server, I guess, and in that goroutine trigger the `wg.Done`.

Comment (zxDiscovery) [reply webhook/webhook.go]: Thanks for your review. @mattmoor @vagababov 
Now I try to find a method to make sure the server is ready to create. Can I use the `http.Get()` method to access the server to check the server is ready?

**Comment by zxDiscovery** [comment]
The CI test pass. Please help to review, thanks! @mattmoor @vagababov @grantr

File: webhook/webhook.go [hunk 3]
```diff
@@ -265,8 +277,113 @@ func configureCerts(ctx context.Context, client kubernetes.Interface, options *C
...
+			case endpointCheckShutdown:
+				return true
+			case endpointCheckReady:
+				return false
+			case endpointCheckNotReady:
+				// continue waiting for endpoint to be ready
+			}
+		}
+	}
+}
```
Comment (mattmoor) [inline webhook/webhook.go]: I'm concerned about the amount of code we are using to solve this (without unit testing).

The initial implementation of this had a deadlock, and this basically uses an Endpoints informer to perform a readiness check, which itself has two problems:
1) Even Endpoints propagation is eventually consistent and no guarantee that the API Server will reach our pod.
2) Our webhooks don't currently define readiness probes, so they will report as ready instantly (before the http endpoint is available).

Let's discuss options more in the issue first?

**Comment by mattmoor** [comment]
Going to close this, as I'm not sure we're in agreement on an approach here.  Let's discuss options in an issue first before jumping into the code.

**Pull request closed without merging**